### PR TITLE
Add PLY Support

### PR DESCRIPTION
**Pull request by sievaxx**
Adds PLY to the Export options. This brings the ability to export Vertex Colours, which OBJ lacks officially.

Models are exported at a 90 degree rotation along the X Axis. I'm not smart enough to figure out how to fix it.

Overall the code is shorter and cleaner then the OBJ exporting code, as well as having the additional feature of vertex colouring. I suggest removing OBJ support entirely in a later revision, but that's up to you.

### CODE REVIEW

File: Math/Convert.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+﻿using System;
+
+
+
+// Yes i really did make a new file for a single function, i want my code to look clean.
```
Comment (figglewatts) [inline Math/Convert.cs]: No need for this comment here :) This is stuff that can be put in a PR description

File: Math/Convert.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+namespace LSDView.Math
+{
+    public class Convert
+    {
+
+        public static byte ToByte(float a)
+        { 
+        return (byte)System.Math.Round(a * 255.0);
```
Comment (figglewatts) [inline Math/Convert.cs]: Incorrect indentation here

File: Util/MeshUtil.cs [hunk 3]
```diff
@@ -248,12 +248,83 @@ public static string RenderableListToObjFile(List<IRenderable> meshes)
...
+            int vertCount = mesh.Verts.Vertices.Length;
+            int triCount = mesh.Verts.Tris;
+
+            PlyBuilder plyString = new PlyBuilder();
+            plyString.WriteHeader(vertCount, triCount);
+
+            // Vertex Elements
+            foreach (var vert in mesh.Verts.Vertices)
+            {
+                plyString.BuildVertexElement(Vector3.TransformPosition(vert.Position, mesh.Transform.Matrix), vert.Normal, vert.TexCoord, vert.Color);
```
Comment (figglewatts) [inline Util/MeshUtil.cs]: Exceeds 120-character line limit as seen in rest of the codebase (sorry this wasn't in the README!)

File: Util/MeshUtil.cs [hunk 4]
```diff
@@ -248,12 +248,83 @@ public static string RenderableListToObjFile(List<IRenderable> meshes)
...
+
+            PlyBuilder plyString = new PlyBuilder();
+            plyString.WriteHeader(vertCount, triCount);
+
+            // Vertex Elements
+            foreach (var mesh in meshes)
+            {
+                foreach (var vert in mesh.Verts.Vertices)
+                {
+                    plyString.BuildVertexElement(Vector3.TransformPosition(vert.Position, mesh.Transform.Matrix), vert.Normal, vert.TexCoord, vert.Color);
```
Comment (figglewatts) [inline Util/MeshUtil.cs]: Also exceeds 120-character line limit

File: Util/MeshUtil.cs [hunk 5]
```diff
@@ -248,12 +248,83 @@ public static string RenderableListToObjFile(List<IRenderable> meshes)
...
+                }
+            }
+
+               //Face Indices
```
Comment (figglewatts) [inline Util/MeshUtil.cs]: Comment is too far over, and needs a space after the `//`

File: Util/PlyBuilder.cs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+namespace LSDView.Util
+{
+    public class PlyBuilder
+    {
+        private StringBuilder _builder;
+
+        public PlyBuilder() { _builder = new StringBuilder(); }
+
+        public PlyBuilder WriteHeader(int verts, int faces)
+            {
```
Comment (figglewatts) [inline Util/PlyBuilder.cs]: Braces indented one level too far

File: Util/PlyBuilder.cs [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+            _builder.AppendLine($"ply\nformat ascii 1.0\n" + // Required header data
+                $"comment Generated by LSDView {Version.String}, created by Figglewatts, 2020\n" + // I have to write the comments here, as the first two lines of any PLY file *MUST* be "ply", followed by the format.
```
Comment (figglewatts) [inline Util/PlyBuilder.cs]: Exceeds 120-char line limit

File: Util/PlyBuilder.cs [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+                $"comment LSDView may be found at https://github.com/Figglewatts/LSDView \n" +
+                $"element vertex {verts}\n" + // Vertex Count
+                $"property float x\nproperty float y\nproperty float z\n" + // Vertex Positions
+                $"property float nx\nproperty float ny\nproperty float nz\n" + // Vertex Normals
+                $"property float s\nproperty float t\n" + // Scale
+                $"property uchar red\nproperty uchar green\nproperty uchar blue\nproperty uchar alpha\n" + // Vertex Colors
```
Comment (figglewatts) [inline Util/PlyBuilder.cs]: Exceeds 120-char line limit

File: Util/PlyBuilder.cs [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+                $"element face {faces}\nproperty list uchar uint vertex_indices\n" + // Face indices
+                $"end_header"); // Yes, this looks messy. This is just how PLY works
+                return this;
+            }
+
+        public PlyBuilder BuildVertexElement(Vector3 pos, Vector3 normal, Vector2 uv, Vector4 col)
+        {
+
+            _builder.AppendLine($"{pos.X} {pos.Y} {pos.Z} {normal.X} {normal.Y} {normal.Z} {uv.X} {uv.Y} {Convert.ToByte(col.X)} {Convert.ToByte(col.Y)} {Convert.ToByte(col.Z)} {Convert.ToByte(col.W)}");
```
Comment (figglewatts) [inline Util/PlyBuilder.cs]: Exceeds 120-char line limit

File: Util/MeshUtil.cs [hunk 10]
```diff
@@ -248,12 +248,83 @@ public static string RenderableListToObjFile(List<IRenderable> meshes)
...
 
+
+        public static string RenderableToPlyFile(IRenderable mesh)
+        {
+            int vertCount = mesh.Verts.Vertices.Length;
+            int triCount = mesh.Verts.Tris;
+
+            PlyBuilder plyString = new PlyBuilder();
+            plyString.WriteHeader(vertCount, triCount);
+
```
Comment (figglewatts) [inline Util/MeshUtil.cs:259]: You can fix the rotation issue with the following code here:
```c#
Matrix4 rotX90Deg = Matrix4.CreateRotationX(MathHelper.DegreesToRadians(-90));

// Vertex elements
foreach (var vert in mesh.Verts.Vertices)
{
    plyString.BuildVertexElement(Vector3.TransformPosition(vert.Position, mesh.Transform.Matrix * rotX90Deg), vert.Normal, vert.TexCoord, vert.Color);
}
```

PLY files must have a slightly different coordinate space, but rotating it by -90 degrees on X seems to fix that :)

File: Util/MeshUtil.cs [hunk 11]
```diff
@@ -248,12 +248,83 @@ public static string RenderableListToObjFile(List<IRenderable> meshes)
...
+            int triCount = 0;
+            foreach (var mesh in meshes)
+            {
+                vertCount += mesh.Verts.Vertices.Length;
+                triCount += mesh.Verts.Tris;
+            }
+
+            PlyBuilder plyString = new PlyBuilder();
+            plyString.WriteHeader(vertCount, triCount);
+
```
Comment (figglewatts) [inline Util/MeshUtil.cs:298]: You can fix the rotation issue with the following code here:
```c#
Matrix4 rotX90Deg = Matrix4.CreateRotationX(MathHelper.DegreesToRadians(-90));

// Vertex elements
foreach (var vert in mesh.Verts.Vertices)
{
    plyString.BuildVertexElement(Vector3.TransformPosition(vert.Position, mesh.Transform.Matrix * rotX90Deg), vert.Normal, vert.TexCoord, vert.Color);
}
```

PLY files must have a slightly different coordinate space, but rotating it by -90 degrees on X seems to fix that :)

**Review by figglewatts — CHANGES_REQUESTED** [review]
Awesome job!! The vertex colouring is really cool, love it.

A few changes requested, mostly formatting stuff and also a suggested fix for the 90 degree rotation problem :)

**Comment by sievaxx** [comment]
Alright, i've fixed everything you commented on.

**Review by figglewatts — APPROVED** [review]
Awesome work! Merging now :)

**Pull request merged**